### PR TITLE
V2 linter API support

### LINKED ISSUE #22: Implement specs
It looks like this linter does not currently have any specs written for it, adding it to the list being maintained [here](https://github.com/AtomLinter/Meta/issues/11) for tracking.

Specs do not need to be super detailed, as they should mainly test whether the package is properly running the linter plugin and parsing its messages. Leave testing of the linter functionality to the author of the linter.

If you need any help feel free to ask!

### LINKED ISSUE #51: Support v2 Linter API
In the [v2.3.0 release](https://github.com/steelbrain/linter/releases/tag/v2.3.0) of `linter` support for the v1 API was removed, breaking semver. This means that this provider is no longer supported by the current version of `linter`.

To get this provider working again it will need to be upgrade to the v2 API, a guide for accomplishing that can be found here: [Upgrading to Standard Linter v2](https://github.com/steelbrain/linter/blob/v2.3.0/docs/guides/upgrading-to-standard-linter-v2.md).

Until this is fixed in order to use this provider you will need to downgrade to the last version of `linter` that supports the v1 API, you can do this with `apm install linter@2.2.0`.

### PR DESCRIPTION
**Pull request by vzamanillo**
Fixes #22.
Fixes #51.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    APM_TEST_PACKAGES: "language-lua"
+  docker:
+    - image: arcanemagus/atom-docker-ci:stable
+  steps:
+    # Restore project state
+    - attach_workspace:
+        at: /tmp
+    - run:
+        name: Install luac and luajit
+        # Note, using ruby-full instead of ruby for the headers to compile rubocop
```
Comment (Arcanemagus) [inline .circleci/config.yml]: 🔥 Not relevant here
```suggestion
```

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        command: |
+          sudo apt-get update && \
+          sudo apt-get install --assume-yes --quiet --no-install-suggests \
+            --no-install-recommends lua5.1 luajit
```
Comment (Arcanemagus) [inline .circleci/config.yml]: Looks like there is a `lua5.3` package, might as well use the current one.
```suggestion
            --no-install-recommends lua5.3 luajit
```

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      env: ATOM_CHANNEL=stable
+
+    - os: linux
+      env: ATOM_CHANNEL=beta
+
+env:
+  global:
+    APM_TEST_PACKAGES: "language-lua"
+
+before_install:
```
Comment (Arcanemagus) [inline .travis.yml]:
```suggestion
install:
```
The `install` step runs before the `script` step, no reason not to just use it here.

File: .travis.yml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  - sudo apt-get update
+  - sudo apt-get install --assume-yes --quiet --no-install-suggests --no-install-recommends lua5.1 luajit
```
Comment (Arcanemagus) [inline .travis.yml]:
```suggestion
  - sudo apt-get install --assume-yes --quiet --no-install-suggests --no-install-recommends lua5.3 luajit
```
Again, might as well use the current version.

File: lib/init.js [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+// eslint-disable-next-line import/no-extraneous-dependencies, import/extensions
+import { CompositeDisposable } from 'atom';
+
+let helpers;
+
+// 1 line
+// 2 message
+// 3 near
+
```
Comment (Arcanemagus) [inline lib/init.js]:
```suggestion
```
Assuming those comments are related to the capturing groups of the regex below.

File: lib/init.js [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+  if (!helpers) {
+    helpers = require('atom-linter');
+  }
+};
+
+const parseLuacOutput = (output, file, editor) => {
+  const messages = [];
+  let match = REGEX.exec(output);
+  while (match !== null) {
+    const line = Number.parseInt(match[1], 10) - 2;
```
Comment (Arcanemagus) [inline lib/init.js]: ❔ It's always **2** off? How odd.

File: lib/init.js [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    match = REGEX.exec(output);
+  }
+  return messages;
+};
+
+module.exports = {
+  activate() {
+    this.idleCallbacks = new Set();
+    let depsCallbackID;
+    const installLinterDummyDeps = () => {
```
Comment (Arcanemagus) [inline lib/init.js]:
```suggestion
    const installLinterLuaDeps = () => {
```

File: lib/init.js [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+      this.idleCallbacks.delete(depsCallbackID);
+      if (!atom.inSpecMode()) {
+        require('atom-package-deps').install('linter-lua');
+      }
+      loadDeps();
+    };
+    depsCallbackID = window.requestIdleCallback(installLinterDummyDeps);
```
Comment (Arcanemagus) [inline lib/init.js]:
```suggestion
    depsCallbackID = window.requestIdleCallback(installLinterLuaDeps);
```

File: lib/init.js [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+  deactivate() {
+    this.idleCallbacks.forEach(callbackID => window.cancelIdleCallback(callbackID));
+    this.idleCallbacks.clear();
+    this.subscriptions.dispose();
+  },
+
+  provideLinter() {
+    return {
+      name: 'luac',
```
Comment (Arcanemagus) [inline lib/init.js]:
```suggestion
      name: 'Lua',
```

File: package.json [hunk 10]
```diff
@@ -1,6 +1,7 @@
 {
   "name": "linter-lua",
-  "main": "./lib/main",
+  "linter-package": true,
```
Comment (Arcanemagus) [inline package.json]:
```suggestion
```
This is an artifact from v0 providers, it shouldn't be in any current package.

File: spec/linter-lua-spec.js [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+  // eslint-disable-next-line no-unused-vars
+  it, fit, wait, beforeEach, afterEach,
+} from 'jasmine-fix';
+
+const { lint } = require('../lib/init.js').provideLinter();
+
+const badFile = path.join(__dirname, 'fixtures', 'bad.lua');
+const goodFile = path.join(__dirname, 'fixtures', 'good.lua');
+
+describe('The dummy provider for Linter', () => {
```
Comment (Arcanemagus) [inline spec/linter-lua-spec.js]:
```suggestion
describe('The Lua provider for Linter', () => {
```

File: spec/linter-lua-spec.js [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+  beforeEach(async () => {
+    atom.workspace.destroyActivePaneItem();
+    await atom.packages.activatePackage('linter-lua');
+  });
+
+  // TODO: Probar también con luajit
```
Comment (Arcanemagus) [inline spec/linter-lua-spec.js]: 🔥 Since it looks like there are tests below for `luajit` this can be removed.

File: .circleci/config.yml (hunk 2 above)
Comment (vzamanillo) [reply .circleci/config.yml]: Seems like lua 5.3 is not available on trusty repo.

**Comment by Arcanemagus** [comment]
Looks like the `lua5.3` package currently has [a bug](https://bugs.launchpad.net/ubuntu/+source/lua5.3/+bug/1707212) with the creation of the `lua` and `luac` aliases.

Add this to the installation step to work around it:
```sh
sudo update-alternatives --install /usr/bin/lua lua-interpreter /usr/bin/lua5.3 130 --slave /usr/share/man/man1/lua.1.gz lua-manual /usr/share/man/man1/lua5.3.1.gz
sudo update-alternatives --install /usr/bin/luac lua-compiler /usr/bin/luac5.3 130 --slave /usr/share/man/man1/luac.1.gz lua-compiler-manual /usr/share/man/man1/luac5.3.1.gz
```

**Comment by vzamanillo** [comment]
> Looks like the `lua5.3` package currently has [a bug](https://bugs.launchpad.net/ubuntu/+source/lua5.3/+bug/1707212) with the creation of the `lua` and `luac` aliases.
> 
> Add this to the installation step to work around it:
> 
> ```shell
> sudo update-alternatives --install /usr/bin/lua lua-interpreter /usr/bin/lua5.3 130 --slave /usr/share/man/man1/lua.1.gz lua-manual /usr/share/man/man1/lua5.3.1.gz
> sudo update-alternatives --install /usr/bin/luac lua-compiler /usr/bin/luac5.3 130 --slave /usr/share/man/man1/luac.1.gz lua-compiler-manual /usr/share/man/man1/luac5.3.1.gz
> ```

Seems like lua 5.3 is not available on trusty repo, the high minor available is 2.

**Comment by Arcanemagus** [comment]
Let's just bump Travis down to 5.2 then. The fixes above should get 5.3 working on Circle though.

**Pull request merged**